src/addagent/validate.c

```diff
@@ -515,19 +523,78 @@ double OS_AgentAntiquity_ID(const char *id) {
     return ret;
 }
 
-/* Returns the number of seconds since last agent connection, or -1 if error. */
+/**
+ * @brief Returns the number of seconds since last agent connection
+ * 
+ * @param name The name of the agent
+ * @param ip The IP address of the agent (unused). Kept only for compatibility
+ * @retval On success, it returns the difference between the current time and the last keepalive
+ * @retval -1 On error: invalid DB query syntax or result
+ */
 double OS_AgentAntiquity(const char *name, const char *ip)
 {
-    struct stat file_stat;
-    char file_name[OS_FLSIZE];
-
-    snprintf(file_name, OS_FLSIZE - 1, "%s/%s-%s", AGENTINFO_DIR, name, ip);
-
-    if (stat(file_name, &file_stat) < 0)
-        return -1;
-
-    return difftime(time(NULL), file_stat.st_mtime);
+    int result = 0;
+    char wdbquery[OS_BUFFER_SIZE] = "";
+    char wdboutput[OS_BUFFER_SIZE] = "";
+    int wdb_sock = -1;
+    time_t output = 0;
+    char *json_string = NULL;
+    cJSON *root = NULL;
+    cJSON *elem = NULL;
+    cJSON *json_name = NULL;
+
+    // Unused
+    (void) ip;
+
+    if(!name){
+        mdebug1("Empty agent name trying to read last keepalive");
+        return OS_INVALID;
+    }
+
+    snprintf(wdbquery, sizeof(wdbquery), SQL_SELECT_KEEPALIVE, name);
+    result = wdbc_query_ex(&wdb_sock, wdbquery, wdboutput, sizeof(wdboutput));
+
+    switch (result){
+        case OS_SUCCESS:
+            break;
+        case OS_INVALID:
+            mdebug1("GLobal DB Error in the response from socket");
+            mdebug2("Global DB SQL query: %s", wdbquery);
+            return OS_INVALID;
+        default:
+            mdebug1("GLobal DB Cannot execute SQL query; err database %s/%s.db", WDB2_DIR, WDB2_GLOB_NAME);
+            mdebug2("Global DB SQL query: %s", wdbquery);
+            return OS_INVALID;
+    }
+
+    if(json_string = wstr_chr(wdboutput, ' '), json_string ){
+        *json_string='\0';
+        json_string++;
+    } else{
+        mdebug1("SQLite result has no space. Query: %s",wdbquery);
+        return OS_INVALID;
+    }
+
+    // The query returned a valid keepalive
+    if(strcmp(wdboutput,"ok") == 0 && strcmp(json_string,"[]") != 0 && strcmp(json_string,"[{}]") != 0){
```

Comment:
Please, make use of the Wazuh DB helper functions located in the same file where `wdbc_query_ex` is implemented to avoid making this JSON parsing here.

Reply:
Done!